src/main/core/logger/shadow_logger.rs

```diff
@@ -0,0 +1,400 @@
+use crate::core::support::simulation_time::SimulationTime;
+use crate::core::worker::Worker;
+use crossbeam::queue::SegQueue;
+use log::{Level, Log, Metadata, Record, SetLoggerError};
+use log_bindings as c_log;
+use once_cell::sync::Lazy;
+use std::cell::RefCell;
+use std::convert::TryFrom;
+use std::sync::mpsc::{Receiver, Sender};
+use std::sync::{Mutex, RwLock};
+use std::time::Duration;
+
+/// Trigger an asynchronous flush when this many lines are queued.
+const ASYNC_FLUSH_QD_LINES_THRESHOLD: usize = 100_000;
+
+/// Performs a *synchronous* flush when this many lines are queued.  i.e. if
+/// after reaching the `ASYNC_FLUSH_QD_LINES_THRESHOLD`, log lines are still
+/// coming in faster than they can actually be flushed, when we reach this limit
+/// we'll pause and let it finish flushing rather than letting the queue
+/// continue growing.
+const SYNC_FLUSH_QD_LINES_THRESHOLD: usize = 10 * ASYNC_FLUSH_QD_LINES_THRESHOLD;
+
+/// Logging thread flushes at least this often.
+const MIN_FLUSH_FREQUENCY: Duration = Duration::from_secs(10);
+
+static SHADOW_LOGGER: Lazy<ShadowLogger> = Lazy::new(|| ShadowLogger::new());
+
+/// Helper for formatting times.
+struct TimeParts {
+    hours: u32,
+    mins: u32,
+    secs: u64,
+    nanos: u64,
+}
+
+impl TimeParts {
+    fn from_nanos(total_nanos: u128) -> Self {
+        let total_secs = u64::try_from(total_nanos / 1_000_000_000).unwrap();
+        let total_mins = u32::try_from(total_secs / 60).unwrap();
+        let hours = total_mins / 60;
+        let mins = total_mins - hours * 60;
+        let secs = total_secs - u64::from(total_mins) * 60;
+        let nanos = u64::try_from(total_nanos - u128::from(total_secs) * 1_000_000_000).unwrap();
```

Comment:
The difference here between `total_*` and just `hours`, `mins`, etc. is unclear. I'm confused why we are converting from nanos to seconds back to nanos; don't we lose information due to rounding? Probably just needs a comment.

Reply:
Added comments and renamed variables to hopefully clarify. Also added a test.